### PR TITLE
feat(flags): Add /secrets resource

### PR DESCRIPTION
**Pull request by cmanallen**
Adds `/secrets` resource which accepts signing-secrets from webhook providers.  Updates the hooks endpoint to verify the signature of a provider payload.

### CODE REVIEW

File: src/sentry/flags/providers.py [hunk 1]
```diff
@@ -127,6 +148,30 @@ def handle_launchdarkly_event(
...
 
+def validate_launchdarkly_event(
+    request_data: bytes,
+    request_headers: HttpHeaders,
+    organization_id: int,
+) -> bool:
+    """Return "true" if the launchdarkly payload is valid."""
+    signature = request_headers.get("X-LD-Signature")
+    if signature is None:
+        return False
```
Comment (aliu39) [inline src/sentry/flags/providers.py:159]: Is auth through a secret always required for LD?

File: src/sentry/flags/providers.py [hunk 2]
```diff
@@ -127,6 +148,30 @@ def handle_launchdarkly_event(
...
+
+    models = FlagWebHookSigningSecretModel.objects.filter(
+        organization_id=organization_id,
+        provider="launchdarkly",
+    ).all()
+    for model in models:
+        if hmac_sha256_hex_digest(model.secret, request_data) == signature:
+            return True
```
Comment (aliu39) [inline src/sentry/flags/providers.py:167]: This check is provider-specific right? Is it expected/common to have multiple secrets?

File: src/sentry/flags/endpoints/secrets.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    authentication_classes = ()
+    owner = ApiOwner.REPLAY
+    permission_classes = ()
+    publish_status = {"POST": ApiPublishStatus.PRIVATE}
+
+    def post(self, request: Request, organization: Organization, provider: str) -> Response:
+        if not features.has(
+            "organizations:feature-flag-audit-log", organization, actor=request.user
+        ):
+            return Response("Not enabled.", status=404)
```
[the remark this answers is not in the document]
Comment (aliu39) [reply src/sentry/flags/endpoints/secrets.py:31]: Scratch this, kinda a nit and I feel like it's easier to miss if in the superclass

File: src/sentry/flags/providers.py (hunk 1 above)
Comment (cmanallen) [reply src/sentry/flags/providers.py:159]: Its not required by LD.  Its required by us.

File: src/sentry/flags/providers.py (hunk 2 above)
Comment (cmanallen) [reply src/sentry/flags/providers.py:167]: Its possible but not common.

File: src/sentry/flags/endpoints/secrets.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+from sentry.api.api_owners import ApiOwner
+from sentry.api.api_publish_status import ApiPublishStatus
+from sentry.api.base import region_silo_endpoint
+from sentry.flags.endpoints import OrganizationFlagsEndpoint
+from sentry.flags.models import FlagWebHookSigningSecretModel
+from sentry.models.organization import Organization
+
+
+class FlagWebhookSigningSecretValidator(serializers.Serializer):
+    secret = serializers.CharField(required=True)
```
[the remark this answers is not in the document]
Comment (cmanallen) [reply src/sentry/flags/endpoints/secrets.py:17]: Secrets can be of any length but will be strings.

File: src/sentry/flags/endpoints/secrets.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            return Response("Not enabled.", status=404)
+
+        validator = FlagWebhookSigningSecretValidator(data=request.data)
+        if not validator.is_valid():
+            return self.respond(validator.errors, status=400)
+
+        FlagWebHookSigningSecretModel.objects.create_or_update(
+            organization=organization,
+            provider=provider,
+            values={
```
[the remark this answers is not in the document]
Comment (cmanallen) [reply src/sentry/flags/endpoints/secrets.py:40]: The first kwargs are query filters.  The values keyword contains the columns to be updated if found.

File: src/sentry/flags/providers.py (hunk 1 above)
Comment (aliu39) [reply src/sentry/flags/providers.py:159]: Line 108 of api.md says "Webhooks are optionally signed by their provider." Maybe we can mention there that we require use of a signature + secret?

File: src/sentry/flags/providers.py (hunk 2 above)
Comment (aliu39) [reply src/sentry/flags/providers.py:167]: Will we support deleting secrets in the future?

**Review by aliu39 — APPROVED** [review]

**Pull request merged**